src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackDecoder.cs

```diff
@@ -0,0 +1,525 @@
+                    if (_integerDecoder.TryDecode(b, out intResult))
+                    {
+                        OnIndexedHeaderName(intResult);
+                    }
+                    break;
+                case State.HeaderNameIndexPostBase:
+                    if (_integerDecoder.TryDecode(b, out intResult))
+                    {
+                        OnIndexedHeaderNamePostBase(intResult);
+                    }
+                    break;
+                case State.HeaderValueLength:
+                    _huffman = (b & HuffmanMask) != 0;
+
+                    // TODO confirm this.
+                    if (_integerDecoder.BeginTryDecode((byte)(b & ~HuffmanMask), StringLengthPrefix, out intResult))
+                    {
+                        OnStringLength(intResult, nextState: State.HeaderValue);
+                        if (intResult == 0)
+                        {
+                            ProcessHeaderValue(handler);
+                        }
+                    }
+                    else
+                    {
+                        _state = State.HeaderValueLengthContinue;
+                    }
+                    break;
+                case State.HeaderValueLengthContinue:
+                    if (_integerDecoder.TryDecode(b, out intResult))
+                    {
+                        OnStringLength(intResult, nextState: State.HeaderValue);
+                        if (intResult == 0)
+                        {
+                            ProcessHeaderValue(handler);
+                        }
+                    }
+                    break;
+                case State.HeaderValue:
+                    _stringOctets[_stringIndex++] = b;
+                    if (_stringIndex == _stringLength)
+                    {
+                        ProcessHeaderValue(handler);
+                    }
+                    break;
+                case State.HeaderFieldIndex:
+                    if (_integerDecoder.TryDecode(b, out intResult))
+                    {
+                        OnIndexedHeaderField(intResult, handler);
+                    }
+                    break;
+                case State.PostBaseIndex:
+                    if (_integerDecoder.TryDecode(b, out intResult))
+                    {
+                        OnPostBaseIndex(intResult, handler);
+                    }
+                    break;
+                case State.LiteralHeaderFieldWithNameReference:
+                    break;
+            }
+        }
+
+        private void OnStringLength(int length, State nextState)
+        {
+            if (length > _stringOctets.Length)
+            {
+                if (length > _maxHeadersLength)
+                {
+                    throw new QPackDecodingException(SR.Format(SR.net_http_headers_exceeded_length, _maxHeadersLength));
+                }
+
+                Pool.Return(_stringOctets, true);
+                _stringOctets = Pool.Rent(length);
```

Comment:
This pattern is dangerous.  If Rent throws, which it could do for a temporary OOM condition or something like that, the array stored in _stringOctets will have already been returned to the pool but will still be referenced by _stringOctets.  Then if the exception causes the Dispose method on this instance to be invoked, it will against call Return _stringOctets, and end up returning the same array instance to the pool again, at which point bad stuff ensues.  It would be better for this to be:
```C#
byte[] toReturn = _stringOctets;
_stringOctets = null;
Pool.Return(toReturn, clearArray: true);
_stringOctets = Pool.Rent(length);
```
or at a minimum:
```C#
Pool.Return(_stringOctets, clearArray: true);
_stringOctets = null; // in case Rent throws
_stringOctets = Pool.Rent(length);
```
etc.